launchdarkly-server-sdk/src/migrations/migrator.rs

```diff
@@ -0,0 +1,317 @@
+// TODO: Remove this when subsequent PRs have added the required implementations.
+#![allow(dead_code)]
+
+use launchdarkly_server_sdk_evaluation::Context;
+
+use crate::{Client, ExecutionOrder, Origin, Stage};
+
+/// An internally used struct to represent the result of a migration operation along with the
+/// origin it was executed against.
+pub struct MigrationResultPair {
+    value: serde_json::Value,
+    origin: Origin,
+}
+
+/// MigrationResult represents the result of a migration operation. If the operation was
+/// successful, the result will contain a pair of values representing the result of the operation
+/// and the origin it was executed against. If the operation failed, the result will contain an
+/// error.
+type MigrationResult = Result<MigrationResultPair, Box<dyn std::error::Error>>;
```

Comment:
```suggestion
/// MigrationResult represents the result of a migration operation. If the operation was
/// successful, the result will contain a [MigrationResultPair]. If the operation failed, the result will contain an error.
type MigrationResult = Result<MigrationResultPair, Box<dyn std::error::Error>>;
```

I think given the docs above on `MigrationResultPair`, it is slightly confusing to be mentioning "the result" twice - easier to just name the two parts of the variant and let people click over. 🤷‍♂️

Reply:
I end up reworking these structures in a subsequent PR. Is it okay if we merge this as is and then we make any future changes there?

The general shape as it stands on my latest local branch.

```rust
/// An internally used struct to represent the result of a migration operation along with the
/// origin it was executed against.
pub struct MigrationOriginResult {
    origin: Origin,
    result: MigrationResult,
}

// /// MigrationResult represents the result of a migration operation. If the operation was
// /// successful, the result will contain a pair of values representing the result of the operation
// /// and the origin it was executed against. If the operation failed, the result will contain an
// /// error.
type MigrationResult = Result<serde_json::Value, String>;
// TODO: Probably want to use a real error type here.

/// A write result contains the operation results against both the authoritative and
/// non-authoritative origins.
///
/// Authoritative writes are always executed first. In the event of a failure, the
/// non-authoritative write will not be executed, resulting in a `None` value in the final
/// MigrationWriteResult.
pub struct MigrationWriteResult {
    authoritative: MigrationOriginResult,
    nonauthoritative: Option<MigrationOriginResult>,
}
```